File: server/excalibur_server/api/v1/files/structures.py
from typing import Literal, Union
from pydantic import BaseModel, model_serializer


class Filelike(BaseModel):
    name: str
    "Name of file"

    fullpath: str
    "Path to the file from the root directory"


class File(Filelike):
    type: Literal["file"] = "file"

    size: int
    "Size of the file in bytes"

    mimetype: str
    "MIME type of the file"


class Directory(Filelike):
    type: Literal["directory"] = "directory"

    items: list[Union[File, "Directory"]] | None = None
    "List of filelike instances in the directory"

File: app/packages/main/src/lib/files/structures.ts
export type ItemType = "file" | "directory";

export interface FileLike {
    /** Name of item */
    name: string;
    /** Path to the item from the root directory */
    fullpath: string;
    /** Type of the item */
    type: ItemType;
}

export interface File extends FileLike {
    type: "file";
    /** Size of the file in bytes */
    size: number;
    /** MIME type of the file */
    mimetype: string;
}

export interface Directory extends FileLike {
    type: "directory";
    /** List of filelike instances in the directory */
    items: (File | Directory)[] | null;
}
